### DISCUSSION TITLE
Request for comments: Remove byte, short and long

### DISCUSSION THREAD

**Discussion by hexagonrecursion** [colobot/colobot#1691] [Ideas]
## Proposal Status

Waiting for sufficient feedback.

## The problem

[Since colobot 0.2.0](https://github.com/colobot/colobot/pull/1253) we have four integer types: `byte`, `short`, `int` and `long`. I do not think this language feature adds anything useful while adding a nontrivial amount of implementation and documentation complexity. I propose we reduce it back to one type. I will retract the proposal if I see a valid use-case for having multiple integer types in CBot.

### [Let's add them just because we can](https://github.com/colobot/colobot/pull/1253#issuecomment-1817820701)

> These new types were not actually "missing", just a tiny bit of code that enabled them

1. This is not a good approach to programming language design. You did not stop to ask "Why should we have this language feature? What purpose does it serve?"
2. [This commit alone](https://github.com/melex750/colobot/commit/c0cdd84e856e1a69a3eaddcd3f790f4e03115775) 29 changed files with 717 additions and 84 deletions. This is not "tiny". And this does not include the necessary documentation.

### [Help people learn CBot, C++ and/or java](https://github.com/colobot/colobot/pull/1253#issuecomment-1817830968)

> Educational purpose. CBot is a language with syntax similar to C++/Java. This design allows players to smooth transition from one language to another. Existence of C++/Java types in CBot smooths down transition further by making them more alike. Also it makes it possible for new players to learn in-game the difference between them.

tldr: There is nothing useful one can learn about the family of integer types from one programming language that would be applicable to a different programming language. Even if integers in both languages happen to have exactly the same syntax and semantics, the only way for me to learn that they do is to read and compare their documentation.

Case study: java has types called `byte`, `short`, `int` and `long` and C++ has types called `signed char`, `short`, `int`, `long` and `long long`, but other than having similar names their type families have very little in common.

1. C++ has unsigned types, Java does not
2. In java `byte` is 8 bit, `short` is 16 bit, `int` is 32 bit and `long` is 64 bit, but in C++ the number of bits in `short`, `int`, `long`, `long long` and possibly `char` depends on OS, CPU architecture, compiler and compiler options.
3. In java integer overflow causes wrap-around, but in C++ signed integer overflow is undefined behaviour. Note: this does not mean "the expression will be evaluated, but its value may be anything". Instead [the behaviour of **the whole program** is undefined and the compiler can do absolutely anything](https://predr.ag/blog/falsehoods-programmers-believe-about-undefined-behavior/).
4. In C++ `unsigned int + long` makes either `long` or `unsigned long` depending on whether `long` is bigger than `int` (which is implementation defined).
5. In C++ `short + short` is `int` due to [integral promotion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion). Java appears to promote `short` to `int` as well, but **CBot does not**. For broader context: in some languages like rust and go mixed-type arithmetic is a syntax error

### [We need multiple integer types to read/write binary files](https://github.com/colobot/colobot/pull/1253#issuecomment-1817830968)

> File I/O operations. To properly load some files you might need a wider palette of types. Particularly when dealing with files in binary mode instead of text mode. You can do it already with read_byte() and write_byte(), but it's harder to use.

No we don't. [Python has just one integer type and it can read and write binary files just fine](https://docs.python.org/3/library/struct.html). If and when we add the ability to convert integers to and from byte strings of different endianness, signedness and number of bits, it would be easier to use a single integer type for everything. Otherwise we face dilemmas like "should read_unsigned_16bit() return `short` or `int` or should we add a new `unsigned short` type?". Using one type also spares the user from all the perils and pitfalls of doing arithmetic with mixed types.

### Sometimes 32 bits is not enough

Why not make CBot `int` be c++ `int64_t` then? Much simpler than adding `long` to CBot.

### Why do many statically typed languages have a whole raft of integer types?

1. To enable frugal use of memory, memory bandwidth and cache.
2. Some C and C++ code casts pointers to and from `char *` to convert data to/from bytes, but this is a bad idea due to [strict aliasing](https://stackoverflow.com/q/23848188/14516046) as well as potential [alignment problems](https://pzemtsov.github.io/2016/11/06/bug-story-alignment-on-x86.html#what-the-standard-says) and endianness issues.

Does this apply to CBot? No:

1) In the current CBot implementation `byte` is not more frugal than `int` because both are [boxed](https://stackoverflow.com/a/25324/14516046) and accessed via virtual methods. Also [CBotVar has like a dozen fields](https://github.com/colobot/colobot/blob/ecc40dab3dead6ef2c96e89c20dced5c5f90bce4/CBot/src/CBot/CBotVar/CBotVar.h#L676-L707). I also do not thinks we should try to make `byte` more frugal because this adds considerable implementation complexity for minimal gain - only CBot code that actually uses `byte` or `short` would benefit. For our use case it is much more productive to optimize our interpreter in a way that benefits common usage scenarios.
2) If and when we implement serialization and deserialization of integers, we should support reading and writing both big and little endian as well as both signed and unsigned - we do not have enough types. We also are not C - we do not have pointers that we could cast to a different type and neither are we Haskell - our instructions can not behave differently depending on the type of the variable that the expression is assigned to. We will have to encode the size, endianness and signedness in the name of the instruction (e.g. `read_i16be()`) regardless of whether we have just one integer type or many

## Size issues

CBot `byte`, `short`, `int` and `long` are wrappers around C++ `signed char`, `short`, `int` and `long`. The number of bits in `short` has depends on how colobot was compiled (same for `int`, `long` and possibly `signed char`). This can cause CBot code that runs fine one machine to break on another machine.

Regardless of whether we keep 4 different integer types or return to having only one, I think a given type should always have the same number of bits (like in java and rust). I see no advantage to CBot having platform-dependent integer sizes.

## Overflow behaviour issues

* CBot `byte` happens to wrap round on overflow (127 + 1 == -128). Integral promotion helps our implementation of `byte` avoid C++ undefined behaviour
* CBot `short` also wraps around (assuming C++ `int` has at least twice the number of bits then C++ `short` to make sure multiplication does not overflow `int`)
* CBot `int` and `long` may appear to wrap around too, but with the current implementation the behaviour is actually undefined - the C++ compiler may assume it never happens. Absolutely anything could happen.

If we do keep `byte`, `short` and `long`, we should make sure that CBot integer overflow behaviour is portable and avoid C++ undefined behaviour.

## Beyond 64 bit

If and when we want to be able to deserialize unsigned 64 bit numbers, I think making `int` a "big integer" (a number without a fixed size that grows as needed) like Python, Ruby, Erlang, tcl and several Lisp dialects do would be more productive than introducing an unsigned 64 bit integer type that is extremely awkward to use in calculations for this one purpose.

## Proposed solution

1. Remove the `byte`, `short` and `long` keywords
2. Gradually remove obsolete the code from the interpreter

## Ping

@melex750 @tomangelo2 @tomaszkax86 What do you think? Also: are there other people I should ping too?

**Comment by tomaszkax86** [comment]
Removal of these types conflicts with features we plan to add to CBot, like:
* Simpler and more compact `struct` value types, like built-in Point
* Inter-bot communication using socket-like functionality by passing arrays and structs directly as payload
* Extended I/O operations to support proper binary reading from and writing to arrays and structs

Comparing CBot's basic types to Python's basic types is kinda weird because those languages have completely different design philosophy. And Python has all the basic types C++ has, albeit you need to use `array` and `struct` modules and their somewhat awkward interfaces rather than native types. Removing extended basic types from CBot would be permanently removing features that cannot be efficiently or reasonably replicated in any other way.

To address some existing issues:
* Size of data types – this is obviously an implementation issue that I've known about and I've been planning to fix but never got time for that. It's not really a bug in the language itself but rather the compiler/interpreter. `long` was added and most likely tested on Linux where it is 64 bits long, fixing this and other types isn't complicated and would be great.
* Overflow and other arithmetic issues – ditto, it's an implementation bug, not a design issue. Most of those can be addressed by rewriting operators and conversions so that all behavior is meaningfully defined. There's no official document specifying CBot and how things work so we may need to write one at some point.
* Endianness issue – I've been working on this on and off since we moved to C++20; we can check system endianness and do byte swaps when necessary. I plan to rewrite I/O and variety of things to make them work properly regardless of endianness.
* Boxing makes any size optimizations pointless – this is because all data is stored in variant-like list types. This is useful for debugging since we have very explicit trees but bad for anything else. But this goes farther than just basic types. Arrays are implemented as linked lists, making access to far elements noticeably slower. A lot of CBot code needs to be rewritten to become optimal. Removing basic types won't fix that in any way.

Everything else is basically a matter of opinion and usage. Most programs won't need extended integer types but this isn't a feature that can be replicated for programs that do need them. As such, the feature should stay but its implementation needs to be fixed.

**Reply by hexagonrecursion** [reply]
> Extended I/O operations to support proper binary reading from and writing to arrays and structs

There are two different roles that serialization/deserialization can serve:
1. Reading/writing arbitrary existing file formats
2. Sending data from CBot to CBot through a binary channel like a socket or a file

Could you please tell me which one you are talking about?

Option 1 requires a lot more types: unsigned versions of each type as well as a way to specify whether a field should be serialized using big endian or little endian representation. Option 2 has no use for multiple sizes of integers unless you plan to severely restrict the number of bytes that can be sent/recieved per game tick.

**Reply by tomaszkax86** [reply]
Both if possible.

Being able to do I/O operations on files as well as send data between bots using structs and arrays would help simplify the code, often making things like communication trivial.

I considered adding unsigned integer types and maybe we'll do that if we decide that's a good idea. As for endianness, this could be a property of the I/O stream or serialization/deserialization method. When you open a file you could specify endianness there rather than mark individual fields in a struct so if you were to read or write a struct or array data would have automatic byte swapping if necessary. We can also add `byteswap()` to CBot if that seems useful.

And I'd say data packets sent between bots should be limited, to something like 1 MiB of data or even less. For most data packets this isn't an issue but if someone wanted to send a lot of data, they'd either have to split it into smaller packets and/or pack it using small integer types.

**Comment by hexagonrecursion** [comment]
Thank you, @tomaszkax86. I understand the idea behind introducing this types now.
